### PR TITLE
Limit initial size of structures

### PR DESCRIPTION
**Pull request by oscarguindzberg**
- Fixes similar problems to the ones fixed on https://github.com/bitcoinj/bitcoinj/pull/1541
- Add unit tests

### CODE REVIEW

File: core/src/main/java/org/bitcoinj/core/TransactionWitness.java [hunk 1]
```diff
@@ -23,28 +23,31 @@
 public class TransactionWitness {
     public static final TransactionWitness EMPTY = new TransactionWitness(0);
 
-    private final byte[][] pushes;
+    private final List<byte[]> pushes;
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/TransactionWitness.java]: I think this refactoring should go in a first (separate) commit.

File: core/src/main/java/org/bitcoinj/core/TransactionWitness.java [hunk 2]
```diff
@@ -69,11 +74,21 @@ public boolean equals(Object o) {
...
+        }
+        return true;
     }
 
     @Override
     public int hashCode() {
-        return Arrays.deepHashCode(pushes);
+        int hashCode = 1;
+        for (byte[] push : pushes) {
+            hashCode = 31*hashCode + (push==null ? 0 : Arrays.hashCode(push));
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/TransactionWitness.java]: Spaces around `*` and `==` missing.

File: core/src/main/java/org/bitcoinj/core/Utils.java [hunk 3]
```diff
@@ -60,6 +60,13 @@
     /** Hex encoding used throughout the framework. Use with HEX.encode(byte[]) or HEX.decode(CharSequence). */
     public static final BaseEncoding HEX = BaseEncoding.base16().lowerCase();
 
+    /**
+     * Max initial size of variable length arrays or ArrayLists.
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/Utils.java]: ...not all ArrayLists, only those that can be sized easily via protocol.

File: core/src/test/java/org/bitcoinj/core/BlockTest.java [hunk 4]
```diff
@@ -294,4 +298,39 @@ public void isBIPs() throws Exception {
...
+        };
+        byte[] serializedBlock = block.bitcoinSerialize();
+        try {
+            UNITTEST.getDefaultSerializer().makeBlock(serializedBlock, serializedBlock.length);
+            fail("We expect ProtocolException with the fixed code and OutOfMemoryError with the buggy code, so this is weird");
+        } catch (ProtocolException e) {
+            //Expected, do nothing
+        }
+    }
+
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/BlockTest.java]: Unnecessary line.

File: core/src/test/java/org/bitcoinj/core/FilteredBlockAndPartialMerkleTreeTests.java [hunk 5]
```diff
@@ -202,4 +204,35 @@ public void serializeDownloadBlockWithWallet() throws Exception {
...
+
+        };
+        byte[] serializedPmt = pmt.bitcoinSerialize();
+        try {
+            new PartialMerkleTree(UNITTEST, serializedPmt, 0);
+            fail("We expect ProtocolException with the fixed code and OutOfMemoryError with the buggy code, so this is weird");
+        } catch (ProtocolException e) {
+            //Expected, do nothing
+        }
+
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/FilteredBlockAndPartialMerkleTreeTests.java]: Unnecessary line.

File: core/src/test/java/org/bitcoinj/core/TransactionTest.java [hunk 6]
```diff
@@ -469,4 +472,106 @@ public void run() {
...
+                        stream.write(push);
+                    }
+
+                    in.getWitness().bitcoinSerializeToStream(stream);
+                }
+            }
+            // lock_time
+            uint32ToByteStreamLE(getLockTime(), stream);
+        }
+
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/TransactionTest.java]: Unnecessary lines.

**Comment by schildbach** [comment]
Thanks for finding these DDoS vectors! Since it's only 3 occurences, maybe list these occurences in the commit comment? Or use one commit per occurence? I added some inline comments.

**Comment by oscarguindzberg** [comment]
Done

**Comment by schildbach** [comment]
Merged. Thanks!

**Pull request merged**